File: modules/multiplexer/zellij.go
package multiplexer

import (
	"context"
	"fmt"
	"os"
	"strings"
	"sync"
	"time"
)

// ZellijBackend implements Backend for the zellij terminal multiplexer.
type ZellijBackend struct {
	mu    sync.RWMutex
	panes map[string]*PaneInfo
}

func NewZellijBackend() *ZellijBackend {
	return &ZellijBackend{
		panes: make(map[string]*PaneInfo),
	}
}

func (z *ZellijBackend) Name() string { return "zellij" }

func (z *ZellijBackend) IsAvailable() bool {
	_, ok := os.LookupEnv("ZELLIJ_SESSION_NAME")
	return ok
}

func (z *ZellijBackend) CreatePane(ctx context.Context, name string, command string) (PaneInfo, error) {
	args := []string{"action", "new-pane", "--name", name, "--", "sh", "-c", command}
	out, err := runCommand(ctx, "zellij", args...)
	if err != nil {
		return PaneInfo{}, fmt.Errorf("zellij create pane: %w", err)
	}

	paneID := parseZellijPaneID(out, name)
	info := PaneInfo{
		ID:        paneID,
		Name:      name,
		State:     PaneStateRunning,
		CreatedAt: time.Now(),
		Backend:   "zellij",
	}

	z.mu.Lock()
	z.panes[paneID] = &info
	z.mu.Unlock()

	return info, nil
}

func (z *ZellijBackend) ClosePane(ctx context.Context, paneID string) error {
	_, err := runCommand(ctx, "zellij", "action", "close-pane", "--pane-id", paneID)
	if err != nil {
		return fmt.Errorf("zellij close pane %s: %w", paneID, err)
	}

	z.mu.Lock()
	if p, ok := z.panes[paneID]; ok {
		p.State = PaneStateExited
		delete(z.panes, paneID)
	}
	z.mu.Unlock()

	return nil
}

func (z *ZellijBackend) SendText(ctx context.Context, paneID string, text string) error {
	_, err := runCommand(ctx, "zellij", "action", "write-chars", "--pane-id", paneID, text)
	if err != nil {
		return fmt.Errorf("zellij send text to %s: %w", paneID, err)
	}
	return nil
}

func (z *ZellijBackend) FocusPane(ctx context.Context, paneID string) error {
	_, err := runCommand(ctx, "zellij", "action", "focus-pane", "--pane-id", paneID)
	if err != nil {
		return fmt.Errorf("zellij focus pane %s: %w", paneID, err)
	}
	return nil
}

func (z *ZellijBackend) CaptureOutput(ctx context.Context, paneID string) (string, error) {
	out, err := runCommand(ctx, "zellij", "action", "dump-screen", "--pane-id", paneID)
	if err != nil {
		return "", fmt.Errorf("zellij capture output %s: %w", paneID, err)
	}
	return out, nil
}

func (z *ZellijBackend) ListPanes(ctx context.Context) ([]PaneInfo, error) {
	z.mu.RLock()
	defer z.mu.RUnlock()

	panes := make([]PaneInfo, 0, len(z.panes))
	for _, p := range z.panes {
		panes = append(panes, *p)
	}
	return panes, nil
}

// parseZellijPaneID extracts a pane identifier from zellij command output.
// Falls back to a generated ID if parsing fails.
func parseZellijPaneID(output string, name string) string {
	output = strings.TrimSpace(output)
	if output != "" {
		// zellij may output pane info in various formats
		for _, line := range strings.Split(output, "\n") {
			line = strings.TrimSpace(line)
			if line != "" {
				return line
			}
		}
	}
	return fmt.Sprintf("zellij-%s-%d", name, time.Now().UnixNano())
}

File: packages/multiplexer/src/zellij.ts
import { PaneState, type PaneInfo } from "./pane.js";
import type { Backend } from "./backend.js";
import { runCommand } from "./backend.js";

/**
 * ZellijBackend implements Backend for the zellij terminal multiplexer.
 */
export class ZellijBackend implements Backend {
  private panes = new Map<string, PaneInfo>();

  /** name returns the backend identifier used for backend selection. */
  name(): string {
    return "zellij";
  }

  /** isAvailable reports whether we are inside a zellij session (ZELLIJ_SESSION_NAME set). */
  isAvailable(): boolean {
    return !!process.env.ZELLIJ_SESSION_NAME;
  }

  /**
   * createPane opens a new zellij pane running the command and records it locally.
   * The pane ID is parsed from zellij output or synthesized on failure.
   */
  async createPane(name: string, command: string): Promise<PaneInfo> {
    const args = ["action", "new-pane", "--name", name, "--", "sh", "-c", command];
    const out = await runCommand("zellij", args);

    const paneID = parseZellijPaneID(out, name);
    const info: PaneInfo = {
      id: paneID,
      name,
      state: PaneState.Running,
      createdAt: new Date(),
      backend: "zellij",
    };

    this.panes.set(paneID, info);
    return info;
  }

  /** closePane closes the pane via zellij action and drops the local record. */
  async closePane(paneID: string): Promise<void> {
    await runCommand("zellij", ["action", "close-pane", "--pane-id", paneID]);
    this.panes.delete(paneID);
  }

  /** sendText writes characters into the pane via zellij write-chars. */
  async sendText(paneID: string, text: string): Promise<void> {
    await runCommand("zellij", ["action", "write-chars", "--pane-id", paneID, text]);
  }

  /** focusPane switches keyboard focus to the named pane via focus-pane. */
  async focusPane(paneID: string): Promise<void> {
    await runCommand("zellij", ["action", "focus-pane", "--pane-id", paneID]);
  }

  /** captureOutput dumps the current screen contents of the pane via dump-screen. */
  async captureOutput(paneID: string): Promise<string> {
    return runCommand("zellij", ["action", "dump-screen", "--pane-id", paneID]);
  }

  /** listPanes returns the locally-tracked panes created by this backend instance. */
  async listPanes(): Promise<PaneInfo[]> {
    return [...this.panes.values()];
  }
}

/**
 * parseZellijPaneID extracts a pane identifier from zellij command output.
 * Falls back to a generated ID if parsing fails.
 */
function parseZellijPaneID(output: string, name: string): string {
  const trimmed = output.trim();
  if (trimmed) {
    for (const line of trimmed.split("\n")) {
      const l = line.trim();
      if (l) return l;
    }
  }
  return `zellij-${name}-${Date.now()}`;
}
